src/hotspot/share/cds/archiveHeapLoader.cpp

```diff
@@ -114,12 +114,36 @@ class PatchCompressedEmbeddedPointers: public BitMapClosure {
     narrowOop* p = _start + offset;
     narrowOop v = *p;
     assert(!CompressedOops::is_null(v), "null oops should have been filtered out at dump time");
-    oop o = ArchiveHeapLoader::decode_from_archive(v);
+    oop o = ArchiveHeapLoader::decode_from_mapped_archive(v);
     RawAccess<IS_NOT_NULL>::oop_store(p, o);
     return true;
   }
 };
 
+class PatchCompressedEmbeddedPointersQuick: public BitMapClosure {
+  narrowOop* _start;
+  uint32_t _delta;
+
+ public:
+  PatchCompressedEmbeddedPointersQuick(narrowOop* start, uint32_t delta) : _start(start), _delta(delta) {}
+
+  bool do_bit(size_t offset) {
+    narrowOop* p = _start + offset;
+    narrowOop v = *p;
+    assert(!CompressedOops::is_null(v), "null oops should have been filtered out at dump time");
+    narrowOop new_v;
+    new_v = static_cast<narrowOop>(static_cast<uint32_t>(v) + _delta);
```

Comment:
I think it is better to use existing APIs for casting to and from `narrowOop`. This statement can be rewritten as:
`narrowOop new_v = CompressedOops::narrow_oop_cast(CompressedOops::narrow_oop_value(v) + _delta)`